### DISCUSSION TITLE
Memory usage before and scopes

### DISCUSSION THREAD

**Discussion by mtungekar** [mjansson/rpmalloc#266] [General]
Wanted to know more about memory usage before and after a scope.  i.e

InitExample();
DoSomeWork();
DeinitExample();

Mainly if after scope exit is memory decommitted and freed ?. In this case memory in use should ideally be approximately same as before. And that seems to be the case when using spRealArrays or simple data types. When loading scene it seems to release most of it. However there is quiet a sizeable amount that is left behind. And if scene is just copied and released this offset grows.  The finalize call does release everything so not a leak. However trying to understand if its caused by a superspans not being release or are there orphaned heaps?

**Comment by mjansson** [comment]
Normally each heap will cache a certain amount of spans (each span is by default 64KiB) defined by the `MAX_THREAD_SPAN_CACHE` define, as well as a number of large spans (up to 4MiB) defined by the `MAX_THREAD_SPAN_LARGE_CACHE` define. You can tweak those to reduce the cache for each heap/thread.

Then the global cache for all heaps/threads is a multiplier of each thread cache size, defined by `GLOBAL_CACHE_MULTIPLIER`.

As long as the allocator is initialized it will hang on to these caches to improve performance, hence calls to `delete/free` will not reduce overall process memory commit size unless caches overflow (or if it's a "huge" allocation >4MiB).

Now, when `rpmalloc_finalize` is called, all caches should be emptied and all memory unmapped by `VirtualFree` so nothing should remain unless there are leaks, in which case the allocator will not properly free those pages. If I recall correctly the wrappings I put in place around rpmalloc should catch any dangling unreleased pointers.
